### PR TITLE
[distGB] graphbolt graph edge's mask will be filled with 0 if these edges have no mask initial

### PR DESCRIPTION
**Pull request by CfromBU**
Add a new parameter "padding" to add_edge_attribute, the edge's mask will be filled with padding rather than 0.
When use DistEdgeDataloader the padding will be set to 1 to fully sample the edge with no mask.

### CODE REVIEW

**Comment by CfromBU** [comment]
1

**Comment by CfromBU** [comment]
@thvasilo ，this pr can fix the graphbolt's mask issue in graphstorm.

**Review by Rhett-Ying — COMMENTED** [review]
add dedicated testcases for the bug

File: tests/distributed/test_distributed_sampling.py [hunk 1]
```diff
@@ -1858,6 +1859,92 @@ def test_local_sampling_heterograph(num_parts, use_graphbolt, prob_or_mask):
                 )
 
 
+def check_mask_hetero_sampling_gb(tmpdir, num_server, use_graphbolt=True):
+    def create_hetero_graph(dense=False, empty=False):
```
Comment (thvasilo) [inline tests/distributed/test_distributed_sampling.py:1863]: I see there's already a function named `create_random_hetero()` could that be used here?

File: tests/distributed/test_distributed_sampling.py [hunk 2]
```diff
@@ -1858,6 +1859,92 @@ def test_local_sampling_heterograph(num_parts, use_graphbolt, prob_or_mask):
...
+    )
+
+    block = next(iter(loader))[2][0]
+    assert block.num_src_nodes("n1") > 0
+
+
+@pytest.mark.parametrize("num_parts", [1])
+def test_local_masked_sampling_heterograph_gb(
+    num_server,
+):
```
Comment (thvasilo) [inline tests/distributed/test_distributed_sampling.py:1941]: If there's only one value do we need to parametrize?

Also, ensure parametrized string name and input argument have the same name

File: tests/distributed/test_distributed_sampling.py (hunk 1 above)
Comment (CfromBU) [reply tests/distributed/test_distributed_sampling.py:1863]: The graph created by create_random_hetero() contains r12, r23, and r13 relations.

When applying a mask to r12, neither DGLGraph nor GB format can sample the n3 node.
When applying a mask to r23, both DGLGraph and GB format can sample the n1 node.
When applying a mask to r13, both DGLGraph and GB format can sample the n2 node.

So the graph can't be used here and I create a new function.

File: tests/distributed/test_distributed_sampling.py (hunk 2 above)
Comment (CfromBU) [reply tests/distributed/test_distributed_sampling.py]: Sorry, the code here was a bit rough, but it has been modified now.

File: tests/distributed/test_distributed_sampling.py (hunk 1 above)
Comment (thvasilo) [reply tests/distributed/test_distributed_sampling.py]: OK. Can we add a comment why this particular graph structure is needed for this test?

Comment (CfromBU) [reply tests/distributed/test_distributed_sampling.py]: Oh, I re-ran the tests and found that the previous statement was incorrect regarding the use of the `create_random_hetero`. Specifically, before the bug fix, `n1` couldn't be sampled when mask was only applied to `r23`, but after the fix, `n1` can now be sampled by `r13`. So, I've switched to using `create_random_hetero` for the tests.

File: python/dgl/distributed/dist_dataloader.py [hunk 3]
```diff
@@ -864,6 +869,7 @@ def __init__(self, g, eids, graph_sampler, device=None, **kwargs):
             else:
                 dataloader_kwargs[k] = v
 
+        collator_kwargs["gb_padding"] = 1
```
Comment (Rhett-Ying) [inline python/dgl/distributed/dist_dataloader.py:872]: this assignment of `1` looks weird. add detailed comment about why we specify it as `1`.

File: python/dgl/distributed/dist_graph.py [hunk 4]
```diff
@@ -143,15 +143,16 @@ def _copy_data_from_shared_mem(name, shape):
 class AddEdgeAttributeFromKVRequest(rpc.Request):
     """Add edge attribute from kvstore to local GraphBolt partition."""
 
-    def __init__(self, name, kv_names):
+    def __init__(self, name, kv_names, padding=0):
```
Comment (Rhett-Ying) [inline python/dgl/distributed/dist_graph.py:146]: this is internal function, let's do not specify default value for it, so we could detect if we miss something. this will enforce us to specify it.

File: python/dgl/distributed/dist_graph.py [hunk 5]
```diff
@@ -1620,13 +1621,15 @@ def _get_edata_names(self, etype=None):
                 edata_names.append(name)
         return edata_names
 
-    def add_edge_attribute(self, name):
+    def add_edge_attribute(self, name, padding=0):
```
Comment (Rhett-Ying) [inline python/dgl/distributed/dist_graph.py:1624]: ditto.

File: python/dgl/distributed/dist_graph.py [hunk 6]
```diff
@@ -169,7 +170,7 @@ def process_request(self, server_state):
             gpb = server_state.partition_book
             # Initialize the edge attribute.
             num_edges = g.total_num_edges
-            attr_data = torch.zeros(num_edges, dtype=data_type)
+            attr_data = torch.full((num_edges,), self._padding, dtype=data_type)
```
Comment (Rhett-Ying) [inline python/dgl/distributed/dist_graph.py:173]: add detailed comment here to explain why we use `padding` and the sampling discrepancy between GB and DGL.

Comment (Rhett-Ying) [inline python/dgl/distributed/dist_graph.py:173]: what if we simply replace `torch.zeros()` with `torch.ones()` without adding `padding` argument?

File: python/dgl/distributed/dist_dataloader.py [hunk 7]
```diff
@@ -327,9 +327,11 @@ def add_edge_attribute_to_graph(g, data_name):
             The graph.
         data_name : str
             The name of data that's stored in DistGraph.ndata/edata.
+        gb_padding : int, optional
+            The padding value for GraphBolt partitions' new edge_attributes.
```
Comment (classicsong) [inline python/dgl/distributed/dist_dataloader.py:331]: Can you add a doc string about what does the default value `0` mean for gb_padding?

Comment (classicsong) [reply python/dgl/distributed/dist_dataloader.py:331]: And what does edge_attributes mean and how it is used?

File: python/dgl/distributed/dist_dataloader.py (hunk 3 above)
Comment (classicsong) [reply python/dgl/distributed/dist_dataloader.py:872]: Agree with Rhett-Ying

Comment (classicsong) [reply python/dgl/distributed/dist_dataloader.py:872]: Another question, why not set the default value of gb_padding to 1?

File: tests/distributed/test_distributed_sampling.py [hunk 8]
```diff
@@ -1858,6 +1858,78 @@ def test_local_sampling_heterograph(num_parts, use_graphbolt, prob_or_mask):
...
+    loader = dgl.dataloading.DistEdgeDataLoader(
+        dist_graph, edges, sampler, batch_size=64
+    )
+    dgl.distributed.exit_client()
+    for p in pserver_list:
+        p.join()
+        assert p.exitcode == 0
+
+    block = next(iter(loader))[2][0]
+    assert block.num_src_nodes("n1") > 0
```
Comment (classicsong) [inline tests/distributed/test_distributed_sampling.py:1921]: Can you also check whether `r12`, `r13` and `r23` edges exists in the blocks? And also check the number of edges?

File: python/dgl/distributed/dist_graph.py (hunk 6 above)
Comment (CfromBU) [reply python/dgl/distributed/dist_graph.py:173]: I am not sure if other features should also be padded with 1, so I use `torch.full` to allow other functions to specify the padding value.

File: python/dgl/distributed/dist_dataloader.py (hunk 3 above)
Comment (CfromBU) [reply python/dgl/distributed/dist_dataloader.py:872]: > Another question, why not set the default value of gb_padding to 1?
Ok, I add a argument gb_padding and set the default value as 1. I previously thought it wasn't necessary to expose this parameter to the user because I wanted it to always be set to 1 to ensure that the sampling works correctly.

File: tests/distributed/test_distributed_sampling.py (hunk 8 above)
Comment (CfromBU) [reply tests/distributed/test_distributed_sampling.py:1921]: Ok, I will check the number of `r12`, `r13`, `r23` in this function.

File: python/dgl/distributed/dist_graph.py (hunk 6 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_graph.py:178]: right. we should not change other data's default values.

File: python/dgl/distributed/dist_dataloader.py (hunk 7 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_dataloader.py:331]: please update the docstring

File: python/dgl/distributed/dist_dataloader.py [hunk 9]
```diff
@@ -508,6 +514,10 @@ class EdgeCollator(Collator):
 
         A set of builtin negative samplers are provided in
         :ref:`the negative sampling module <api-dataloading-negative-sampling>`.
+    gb_padding : int, optional
+        The padding value for GraphBolt partitions' new edge_attributes if the attributes in DistGraph are None.
+        e.g. prob/mask-based sampling.
+        Only when the mask of one edge is set as 1, the edge will be sampled.
```
Comment (Rhett-Ying) [inline python/dgl/distributed/dist_dataloader.py:520]: more details about why we set default value as 1 while set 0 for node counterpart

File: python/dgl/distributed/dist_graph.py [hunk 10]
```diff
@@ -169,7 +170,12 @@ def process_request(self, server_state):
             gpb = server_state.partition_book
             # Initialize the edge attribute.
             num_edges = g.total_num_edges
-            attr_data = torch.zeros(num_edges, dtype=data_type)
+
+            # Padding is used here to fill missing edge attributes (e.g., 'prob' or 'mask') for certain edge types.
+            # In DGLGraph, some edges may not have attributes or their values could be None.
```
Comment (Rhett-Ying) [inline python/dgl/distributed/dist_graph.py:175]: not clear enough. we should make it clear that  the None edges will be sampled in DGL, not sampled while in GB.

**Review by Rhett-Ying — APPROVED** [review]
basically LGTM. please resolve all threads after changes. and always resolve all threads before merge. Make sure the code works well in the case we hit the issue, and unit test passes in your local as CI is down for now.

File: python/dgl/distributed/dist_dataloader.py [hunk 11]
```diff
@@ -344,6 +346,10 @@ class NodeCollator(Collator):
         The node set to compute outputs.
     graph_sampler : dgl.dataloading.BlockSampler
         The neighborhood sampler.
+    gb_padding : int, optional
+        The padding value for GraphBolt partitions' new edge_attributes.
+        e.g. some edges of specific types have no mask, the mask will be set as gb_padding.
+        the edge will not be sampled if the mask is 0.
```
Comment (classicsong) [inline python/dgl/distributed/dist_dataloader.py:352]:
```suggestion
        An edge will not be sampled if the mask is 0.
```

**Review by classicsong — APPROVED** [review]
LGTM

File: python/dgl/distributed/dist_graph.py (hunk 10 above)
Comment (CfromBU) [reply python/dgl/distributed/dist_graph.py]: Ok, I have enriched the comment and explained the None edges will be sampled in DGLGraph.

File: python/dgl/distributed/dist_dataloader.py (hunk 11 above)
Comment (CfromBU) [reply python/dgl/distributed/dist_dataloader.py]: OK, I have changed the comment.

File: python/dgl/distributed/dist_dataloader.py (hunk 9 above)
Comment (CfromBU) [reply python/dgl/distributed/dist_dataloader.py]: I have set the gb_padding as 1 in NodeCollator and changed the comment correspondingly.

**Comment by CfromBU** [comment]
图机器学习算法库作业已收到！
